packages/@lwc/engine-core/src/framework/api.ts

```diff
@@ -466,20 +468,25 @@ function i(
  * [f]lattening
  * @param items
  */
-function f(items: Readonly<Array<Readonly<Array<VNodes>> | VNodes>>): VNodes {
+function f(items: ReadonlyArray<ReadonlyArray<VNodes> | VNodes>): VNodes {
     if (process.env.NODE_ENV !== 'production') {
         assert.isTrue(isArray(items), 'flattening api can only work with arrays.');
     }
     const len = items.length;
-    const flattened: VNodes = [];
+    const flattened: MutableVNodes = [];
     // TODO [#1276]: compiler should give us some sort of indicator when a vnodes collection is dynamic
     sc(flattened);
     for (let j = 0; j < len; j += 1) {
         const item = items[j];
-        if (isArray(item)) {
-            ArrayPush.apply(flattened, item);
-        } else {
-            ArrayPush.call(flattened, item);
+        const innerLen = items[j].length;
+        for (let k = 0; k < innerLen; k += 1) {
+            const vnodeOrArray = item[k];
+            if (isArray(vnodeOrArray)) {
+                ArrayPush.apply(flattened, vnodeOrArray);
+            } else {
+                // `isArray` doesn't narrow this block properly...
+                ArrayPush.call(flattened, vnodeOrArray as VNode | null);
+            }
```

Comment:
Most of the changes in this PR are just type changes (and primarily dealing with `ArraySlice.call(arguments)`). This is a JS change, which is a notable exception, because it's a bug! 🐛 

We were trying to flatten a weirdly nested array of `VNodes`, which is itself an array type, so the `isArray` check wasn't doing the flattening we expected. This fixes that so that the result actually is `VNodes`, rather than `Array<VNode | VNodes>` (I think that's what it was? Too much nesting!)

Reply:
This new logic looks like it may have a perf impact since we are now traversing more deeply than we would have otherwise.

Are you able to write a Karma test to reproduce the bug and demonstrate the fix? It's possible that the structure of the array is guaranteed by the compiler, so this isn't actually deeply nested in practice.

Reply:
Upon further consideration, I would guess that it's actually the function signature is incorrect, rather than the implementation. Just an extra wrapping of `ReadonlyArray<...>` that isn't actually what it wants.